File: src/main/java/com/gmail/borlandlp/minigamesdtools/util/NMSUtil.java
package com.gmail.borlandlp.minigamesdtools.util;

import net.minecraft.server.v1_12_R1.BiomeBase;
import net.minecraft.server.v1_12_R1.Entity;
import net.minecraft.server.v1_12_R1.EntityTypes;
import net.minecraft.server.v1_12_R1.MinecraftKey;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/*
* ToDo: Отказаться от рефлексии для использования кастомного поведения для энтити?
* */
public final class NMSUtil {
    public static void registerEntity(final String name, final int id, final Class<? extends Entity> nmsClass,
                                      final Class<? extends Entity> customClass) {
        try {
            final List<Map<?, ?>> dataMaps = new ArrayList<Map<?, ?>>();
            for (Field f : EntityTypes.class.getDeclaredFields()) {
                if (f.getType().equals(Map.class)) {
                    f.setAccessible(true);
                    dataMaps.add((Map<?, ?>) f.get(null));
                }
            }
            EntityTypes.b.a(id, new MinecraftKey(name), customClass);
            for (Field f : BiomeBase.class.getDeclaredFields()) {
                /*
                 * Viva64 report
                 * https://www.viva64.com/ru/w/v6054/
                 * */
                //if (f.getType().getSimpleName().equals(Map.class.getSimpleName())) {
                if (f.getType().equals(BiomeBase.class)) {
                    if (f.get(null) != null) {
                        for (Field list : BiomeBase.class.getDeclaredFields()) {
                            if (list.getType().equals(List.class)) {
                                list.setAccessible(true);
                                @SuppressWarnings("unchecked")
                                List<BiomeBase.BiomeMeta> metaList = (List<BiomeBase.BiomeMeta>) list.get(f.get(null));

                                for (BiomeBase.BiomeMeta meta : metaList) {
                                    Field clazz = BiomeBase.BiomeMeta.class.getDeclaredFields()[0];
                                    if (clazz.get(meta).equals(nmsClass))
                                        clazz.set(meta, customClass);
                                }
                            }
                        }
                    }
                }
            }
        } catch (final Exception e) {
            e.printStackTrace();
        }
    }

    public static void unregisterEntity(final String name, final int id, final Class<? extends Entity> nmsClass,
                                      final Class<? extends Entity> customClass) {
        try {
            final List<Map<?, ?>> dataMaps = new ArrayList<Map<?, ?>>();
            for (Field f : EntityTypes.class.getDeclaredFields()) {
                /*
                * Viva64 report
                * https://www.viva64.com/ru/w/v6054/
                * */
                //if (f.getType().getSimpleName().equals(Map.class.getSimpleName())) {
                if (f.getType().equals(Map.class)) {
                    f.setAccessible(true);
                    dataMaps.add((Map<?, ?>) f.get(null));
                }
            }
            EntityTypes.b.a(id, new MinecraftKey(name), customClass);
            for (Field f : BiomeBase.class.getDeclaredFields()) {
                if (f.getType().equals(BiomeBase.class)) {
                    if (f.get(null) != null) {
                        for (Field list : BiomeBase.class.getDeclaredFields()) {
                            if (list.getType().equals(List.class)) {
                                list.setAccessible(true);
                                @SuppressWarnings("unchecked")
                                List<BiomeBase.BiomeMeta> metaList = (List<BiomeBase.BiomeMeta>) list.get(f.get(null));

                                for (BiomeBase.BiomeMeta meta : metaList) {
                                    Field clazz = BiomeBase.BiomeMeta.class.getDeclaredFields()[0];
                                    if (clazz.get(meta).equals(customClass))
                                        clazz.set(meta, nmsClass);
                                }
                            }
                        }
                    }
                }
            }
        } catch (final Exception e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/util/NMSUtil.kt
package com.gmail.borlandlp.minigamesdtools.util

import net.minecraft.server.v1_12_R1.BiomeBase
import net.minecraft.server.v1_12_R1.BiomeBase.BiomeMeta
import net.minecraft.server.v1_12_R1.Entity
import net.minecraft.server.v1_12_R1.EntityTypes
import net.minecraft.server.v1_12_R1.MinecraftKey
import java.util.*

/*
* ToDo: Отказаться от рефлексии для использования кастомного поведения для энтити?
* */
object NMSUtil {
    @JvmStatic
    fun registerEntity(
        name: String?, id: Int, nmsClass: Class<out Entity?>,
        customClass: Class<out Entity?>?
    ) {
        try {
            val dataMaps: MutableList<Map<*, *>> =
                ArrayList()
            for (f in EntityTypes::class.java.declaredFields) {
                if (f.type == MutableMap::class.java) {
                    f.isAccessible = true
                    dataMaps.add(f[null] as Map<*, *>)
                }
            }
            EntityTypes.b.a(id, MinecraftKey(name), customClass)
            for (f in BiomeBase::class.java.declaredFields) {
                if (f.type == BiomeBase::class.java) {
                    if (f[null] != null) {
                        for (list in BiomeBase::class.java.declaredFields) {
                            if (list.type == MutableList::class.java) {
                                list.isAccessible = true
                                val metaList =
                                    list[f[null]] as List<BiomeMeta>
                                for (meta in metaList) {
                                    val clazz =
                                        BiomeMeta::class.java.declaredFields[0]
                                    if (clazz[meta] == nmsClass) clazz[meta] = customClass
                                }
                            }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun unregisterEntity(
        name: String?, id: Int, nmsClass: Class<out Entity?>?,
        customClass: Class<out Entity?>
    ) {
        try {
            val dataMaps: MutableList<Map<*, *>> =
                ArrayList()
            for (f in EntityTypes::class.java.declaredFields) { /*
                * Viva64 report
                * https://www.viva64.com/ru/w/v6054/
                * */
//if (f.getType().getSimpleName().equals(Map.class.getSimpleName())) {
                if (f.type == MutableMap::class.java) {
                    f.isAccessible = true
                    dataMaps.add(f[null] as Map<*, *>)
                }
            }
            EntityTypes.b.a(id, MinecraftKey(name), customClass)
            for (f in BiomeBase::class.java.declaredFields) {
                if (f.type == BiomeBase::class.java) {
                    if (f[null] != null) {
                        for (list in BiomeBase::class.java.declaredFields) {
                            if (list.type == MutableList::class.java) {
                                list.isAccessible = true
                                val metaList =
                                    list[f[null]] as List<BiomeMeta>
                                for (meta in metaList) {
                                    val clazz =
                                        BiomeMeta::class.java.declaredFields[0]
                                    if (clazz[meta] == customClass) clazz[meta] = nmsClass
                                }
                            }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}